runtime/java/src/common/Debug.java

```diff
@@ -0,0 +1,1198 @@
+        this.toggleHeadlessAttributes = false;
+        this.toggleChoices = new String[] {"nameDisplay", "cStackDisplay", "fullAttributeNames"};
+
+        //A few Place holder variable
+        DecoratedNode childNode;
+        this.root = tree;
+        this.currentNode = tree;
+
+        //This stack is used for the undo command
+        this.nodeStack = new Stack<DecoratedNode>();
+
+        // These stacks are important for the file visualization
+        this.cStack = new FileContextVisualization("context.txt", "********************************");
+        cStack.push(currentNode);
+        this.contextStack = (ContextStack)cStack.getContextStack();
+        this.sStack = new SimplifiedContextStack(contextStack);
+        sStack.generateHTMLFile();
+
+        //Start of display
+        System.out.println("Enter characters, and 'q' to quit.");
+        if(toggleCStackDisplay){
+            cStack.show();
+        }
+        if(toggleNameDisplay){
+            printName(currentNode);
+        }
+
+        //Main Control loop
+        loop: do { 
+            //After each command is done the user is prompted again
+            System.out.print(">DEBUGGER-PROMPT$");
+            userInput = inp.nextLine();
+            userInputList = userInput.split(" ");
+
+            //The basic structure is we check if the user input matches any expected input
+            //If so we call the corresponding function where the work is done
+            //Most IO is done in this function and most logic is done in the auxiliary functions
+            switch (userInputList[0]) {
+
+                //used for going to a parent of the current node
+                case "up": case "u": 
+
+                    //up should not have any arguments
+                    if (userInputList.length != 1) {
+                        System.out.println("invalid, correct usage: up");
+                        break;
+                    }
+
+                    //If so we can go to that node
+                    if (up() == -1)
+                        break;
+
+                    //Update the json and html and terminal to the new node
+                    updateDisplay();
+                    break;
+
+                //Used to navigate to children
+                case "down": case "d":  
+                    String childName = "";
+
+                    //If the user does not provide a child we should list them out
+                    if (userInputList.length == 1) {
+                        String currentProduction = currentNode.undecorate().getProdleton().getTypeUnparse();
+                        String[] listCurrentProduction = currentProduction.split("\\s+");
+                        String[] childNames = Arrays.copyOfRange(listCurrentProduction, 2, listCurrentProduction.length);
+                        System.out.println("Which child?");
+                        displayArray(childNames);
+                        System.out.print(">DEBUGGER-PROMPT$");
```

Comment:
It's a bit confusing that the prompt is the same for choosing the child. Could just have "Which child? " as the prompt.